callflow/modules/histogram.py

```diff
@@ -9,81 +9,95 @@
 class Histogram:
     """
     Calculate Histogram (Per rank, All ranks, Per dataset)
     """
 
-    HISTO_TYPES = ["rank", "name", "dataset", "all_ranks"]
-    KEYS_AND_ATTRS = {"Inclusive": "time (inc)", "Exclusive": "time"}
+    HISTO_TYPES = ["rank", "name", "dataset"]
 
-    def __init__(self, df_ensemble, df_target=None, bins=20):
+    def __init__(self, dataframe, relative_to_df=None, bins=20,
+                 histo_types = [], proxy_columns={}):
         """
 
-        :param df_ensemble:
-        :param df_target:
+        :param dataframe:
+        :param histo_types:
         :param bins:
+        :param proxy_columns:
         """
-        assert isinstance(df_ensemble, pd.DataFrame)
-        assert isinstance(df_target, pd.DataFrame) or df_target is None
+        assert isinstance(dataframe, pd.DataFrame)
+        if relative_to_df is not None:
+            assert isinstance(relative_to_df, pd.DataFrame)
         assert isinstance(bins, int)
+        assert isinstance(proxy_columns, dict)
+        assert isinstance(histo_types, list)
         assert bins > 0
 
-        self.result = {"Inclusive": {}, "Exclusive": {}}
-
-        # for each property
-        for prop in Histogram.HISTO_TYPES:
+        self.time_columns = [proxy_columns.get(_, _) for _ in TIME_COLUMNS]
+        self.result = {_: {} for _ in TIME_COLUMNS}
 
-            if df_target is None:
-                _dfe = Histogram._get_data_by_property(df_ensemble, prop)
-                for k, a in Histogram.KEYS_AND_ATTRS.items():
-                    _hist = histogram(_dfe[a], bins=bins)
-                    self.result[k][prop] = {"ensemble": Histogram._format_data(_hist)}
+        if len(histo_types) == 0:
+            histo_types = Histogram.HISTO_TYPES
 
-            else:
-                _dfe = Histogram._get_data_by_property(df_ensemble, prop)
-                _dft = Histogram._get_data_by_property(df_target, prop)
+        # for each type of histogram and each time column
+        for h,(tk,tv) in itertools.product(histo_types, zip(TIME_COLUMNS, self.time_columns)):
 
-                for k, a in Histogram.KEYS_AND_ATTRS.items():
-                    _e = _dfe[a]
-                    _t = _dft[a]
+            # compute the range of the actual data
+            df = self._get_data_by_histo_type(dataframe, h)[tv]
+            drng = [df.min(), df.max()]
 
-                    _min = min(_e.min(), _t.min())
-                    _max = max(_e.max(), _t.max())
-
-                    _histe = histogram(_e, [_min, _max], bins=bins)
-                    _histt = histogram(_t, [_min, _max], bins=bins)
+            # compute the range of the relative_to_df
+            if relative_to_df is None:
+                rrng = drng
+            else:
+                rdf = self._get_data_by_histo_type(relative_to_df, h)[tv]
+                rrng = [rdf.min(), rdf.max()]
+                assert rrng[0] <= drng[0]
```

Comment:
Harsh, correct me if I am wrong. But I feel we should technically avoid this assertion because say we want to plot the histogram for any other `relative_to_df == ensemble_df`. This statement would fail. 

I checked that, even for `ensemble` scenario, this assertion is triggered (which is also surprising). There might be a bug somewhere.